src/plugins/intel_gpu/src/runtime/memory_pool.cpp

```diff
@@ -272,7 +272,7 @@ memory::ptr memory_pool::get_memory(const layout& layout,
     }
     if (do_reuse) {
         // reusable within the same network
-        if (!layout.format.is_image() && !layout.data_padding) {
+        if (!layout.format.is_image() && (!layout.data_padding || is_dynamic)) {
```

Comment:
Not directly related to your change, but what about reorganizing the code like this for readability?
```
if (layout.format.is_image() {
            // images (reuse not yet implemented)
            auto mem = alloc_memory(layout, type, reset);
} else if (!layout.data_padding || is_dynamic))
            // non-padded buffers. For dynamic shape, use non-padded pool even if it has padding because we will reset the buffer if it is reused
            return get_from_non_padded_pool(layout, prim_id, unique_id, network_id, restrictions, type, reset, is_dynamic);
} else {
            // padded buffers
            return get_from_padded_pool(layout, prim_id, unique_id, network_id, restrictions, type);
}
```

Reply:
applied it